lib/Support/PrettyPrinter.cpp

```diff
@@ -0,0 +1,297 @@
+// on this algorithm.
+//
+// Implementations of note that have interesting modifications for their
+// languages and modernization of the paper's algorithm:
+// * prettyplease / rustc_ast_pretty
+//   Pretty-printers for rust, the first being useful for rustfmt-like output.
+//   These have largely the same code and were based on one another.
+//     https://github.com/dtolnay/prettyplease
+//     https://github.com/rust-lang/rust/tree/master/compiler/rustc_ast_pretty
+//   This is closest to the paper's algorithm with modernizations,
+//   and most of the initial tweaks have also been implemented here (thanks!).
+// * swift-format: https://github.com/apple/swift-format/
+//
+// If we want fancier output or need to handle more complicated constructs,
+// both are good references for lessons and ideas.
+//
+// FWIW, at time of writing these have compatible licensing (Apache 2.0).
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/PrettyPrinter.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace circt {
+namespace pretty {
+
+/// Destructor, anchor.
+PrettyPrinter::Listener::~Listener() = default;
+
+/// Add token for printing.  In Oppen, this is "scan".
+void PrettyPrinter::add(Token t) {
+  // Add token to tokens, and add its index to scanStack.
+  auto addScanToken = [&](auto offset) {
+    auto right = tokenOffset + tokens.size();
+    scanStack.push_back(right);
+    tokens.push_back({t, offset});
+  };
+  llvm::TypeSwitch<Token *, void>(&t)
+      .Case([&](StringToken *s) {
+        // If nothing on stack, directly print
+        FormattedToken f{t, (int32_t)s->text().size()};
+        // Empty string token isn't /wrong/ but can have unintended effect.
+        assert(!s->text().empty() && "empty string token");
+        if (scanStack.empty())
+          return print(f);
+        tokens.push_back(f);
+        rightTotal += f.size;
+        assert(rightTotal > 0);
+        checkStream();
+      })
+      .Case([&](BreakToken *b) {
+        if (scanStack.empty())
+          clear();
+        else
+          checkStack();
+        addScanToken(-rightTotal);
+        rightTotal += b->spaces();
+        assert(rightTotal > 0);
+      })
+      .Case([&](BeginToken *b) {
+        if (scanStack.empty())
+          clear();
+        addScanToken(-rightTotal);
+      })
+      .Case([&](EndToken *end) {
+        if (scanStack.empty())
+          return print({t, 0});
+        addScanToken(-1);
+      });
+  rebaseIfNeeded();
+}
+
+void PrettyPrinter::rebaseIfNeeded() {
+  // Check for too-large totals, reset.
+  // This can happen if we have an open group and emit
+  // many tokens, especially newlines which have artificial size.
+  if (tokens.empty())
+    return;
+  assert(leftTotal >= 0);
+  assert(rightTotal >= 0);
+  if (uint32_t(leftTotal) > ~uint32_t{0} >> 4) {
```

Comment:
How did you determine the parameter `~uint32_t{0} >> 4`?

Reply:
Not as principled as I'd like :slightly_frowning_face:, thanks for pushing/asking.

`uint32_t` is because `leftTotal` is 32bits, 1st bit is because it's signed, next 3 are just because I made up using a limit of 1/8th the maximum (not lower, don't want to rebase too often), but I wanted to revisit.  I had this `1U << 29` earlier IIRC, maybe that makes more sense.  This is not encountered until `neverbreak` bits added (`kInfinity` taking 15bits contributes quite a bit), and with some careful thinking/design I suspect we can avoid this in non-pathological (2GB strings) cases.

We really need to keep `rightTotal` from hitting the limit (it'll go first), but this checks `leftTotal` as rebase only guarantees `leftTotal` is reset-- rightTotal might just be because we have a silly amount of (long?) strings pending.

Reply:
Refactored to a constant and hopefully made its derivation a little clearer.